makefiles/Base.toml

```diff
@@ -7,26 +7,61 @@ CARGO_PROFILE = "release"
 [tasks.set-image-tag]
 env = { IMAGE_TAG = "${CARGO_MAKE_GIT_HEAD_LAST_COMMIT_HASH}-${CARGO_PROFILE}" }
 
-[env]
-FLUVIO_VERSION = { script = ["cat VERSION"] }
+################################################################################
+# Tasks for building targets matching the build host
+#
 
-[tasks.build-release.mac]
+[tasks.build-dev.mac]
+extend = "build-dev-mac"
+[tasks.build-dev.linux]
+extend = "build-dev-musl"
+
+[tasks.build-prod.mac]
+extend = "build-prod-mac"
+[tasks.build-prod.linux]
+extend = "build-prod-musl"
+
+
+################################################################################
+# Building for x86_64-apple-darwin can only be done on Mac
+#
+
+[tasks.build-dev-mac.mac]
```

Comment:
I'm a bit confused by the task named `build-dev-mac` and `build-prod-mac`, what's wrong with just `build-dev.mac` and `build-prod.mac` as well as their linux counter parts? After fighting with the cross compilation tooling, I think it's easier and more maintainable to just compile it for the host itself.

Reply:
Sure. I think of this as two "levels" of tasks. At the lower level, you have tasks that always produce the same thing, and are not influenced by the environment they are run in. These would be things like

```
build-dev-mac
build-prod-mac
build-dev-musl
build-prod-musl
```

By looking at these, the goal is to be able to know _exactly_ what the output artifact is. For example, if I want to build for `x86_64-apple-darwin` in release mode, I'd use `build-prod-mac`, no matter what host I am using.

Then at a higher level, there are tasks that _do_ adapt to the environment in which they are run. The examples of these are

```
build-dev
build-prod
```

The implicit information here is that the build artifact coming out will have a target that matches the host that you are building on. So rather than saying "build a debug image for target X", you are saying "build a debug image for my host". This is essentially the same pattern that cargo uses, if you give it no `--target`, it will build for your host. If you give it a `--target`, it will build exactly that target, regardless of the host.

---

Previously, we _only_ had "high-level" tasks like `build-dev` and `build-prod`. I found this to be a problem in many cases where I wanted to quickly make a build for `musl` from my mac machine, and there was no way to express that with the tasks we had.

Reply:
I think we should not put configuration specific tasks.   This leads to combinational explosion.  For example, if we add another target: `raspberry`.   We should only have 1 single build task which can take parameters as needed

Reply:
I guess the moral of the story is that the low-level tasks are very useful for local development, and the high-level tasks make it easy to make CI jobs that do the right thing per platform.